### PR TITLE
Mttl 1045

### PR DESCRIPTION
**Pull request by humulla**
## Link to JIRA ticket

[Jira Ticket](https://hackney.atlassian.net/jira/software/projects/MTTL/boards/47?selectedIssue=MTTL-1045)

## Describe this PR

Post-tenure endpoint has been created with all related unit tests and E2E tests

#### _Checklist_

- [x] Added end-to-end (i.e. integration) tests where necessary e.g to test the complete functionality of a newly added feature
- [x] Added tests to cover all new production code
- [x] Checked all code for possible refactoring
- [x] Code pipeline builds correctly

### CODE REVIEW

File: TenureInformationApi/V1/Factories/ResponseFactory.cs [hunk 1]
```diff
@@ -36,10 +36,35 @@ public static TenureResponseObject ToResponse(this TenureInformation domain)
...
+                PaymentReference = domain.PaymentReference,
+                PotentialEndDate = domain.PotentialEndDate,
+                StartOfTenureDate = domain.StartOfTenureDate,
+                SubletEndDate = domain.SubletEndDate,
+                SuccessionDate = domain.SuccessionDate,
+                TenuredAsset = domain.TenuredAsset,
+                TenureType = domain.TenureType,
+                Terminated = domain.Terminated,
+                AgreementType = domain.AgreementType
+            }).ToList();
```
Comment (mikemoody-amido) [inline TenureInformationApi/V1/Factories/ResponseFactory.cs:65]: Not sure why you changed this and replicated the entire single-object implementation here.
Couldn't you have just changed it to...
```
return domainList.Select(domain => ToResponse(domain)).ToList();
```

File: TenureInformationApi/V1/Factories/ResponseFactory.cs [hunk 2]
```diff
@@ -5,9 +5,9 @@
 
 namespace TenureInformationApi.V1.Factories
 {
-    public static class ResponseFactory
+    public class ResponseFactory : IResponseFactory
```
Comment (mikemoody-amido) [inline TenureInformationApi/V1/Factories/ResponseFactory.cs:8]: Was there a reason you changed this class from a static one to one injected by the DI container?

File: TenureInformationApi/V1/Boundary/Requests/CreateTenureRequestObject.cs [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    {
+        public Guid Id { get; set; }
+        public string PaymentReference { get; set; }
+        public List<HouseholdMembers> HouseholdMembers { get; set; }
+        public TenuredAsset TenuredAsset { get; set; }
+        public Charges Charges { get; set; }
+        public DateTime StartOfTenureDate { get; set; }
+        public DateTime? EndOfTenureDate { get; set; }
+        public TenureType TenureType { get; set; }
+        public bool IsActive { get; set; }
```
Comment (mikemoody-amido) [inline TenureInformationApi/V1/Boundary/Requests/CreateTenureRequestObject.cs:19]: Should `IsActive` be here (and in swagger) given that it is a field calculated from the `EndOfTenureDate` value?

File: TenureInformationApi/V1/Boundary/Requests/CreateTenureRequestObject.cs [hunk 4]
```diff
@@ -0,0 +1,33 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using TenureInformationApi.V1.Domain;
+
+namespace TenureInformationApi.V1.Boundary.Requests
+{
+    public class CreateTenureRequestObject
```
Comment (mikemoody-amido) [inline TenureInformationApi/V1/Boundary/Requests/CreateTenureRequestObject.cs:9]: There is currently no validation on any of the request object data, but should there be?

File: TenureInformationApi.Tests/V1/UseCase/GetByIdUseCaseTests.cs [hunk 5]
```diff
@@ -19,10 +20,13 @@ public class GetByIdUseCaseTests
         private readonly Mock<ITenureGateway> _mockGateway;
         private readonly GetByIdUseCase _classUnderTest;
         private readonly Fixture _fixture = new Fixture();
+        private readonly Mock<IResponseFactory> _mockResponseFactory;
```
Comment (mikemoody-amido) [inline TenureInformationApi.Tests/V1/UseCase/GetByIdUseCaseTests.cs:23]: You don't actually need a mock here, you could use a real instance.

File: TenureInformationApi.Tests/V1/E2ETests/Stories/PostTenureTests.cs [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+        }
+
+        [Fact]
+        public void ServiceReturnsTheRequestedPerson()
+        {
+            this.Given(g => _tenureFixture.GivenNewTenureRequest())
+                .When(w => _steps.WhenCreateTenureApiIsCalled(_tenureFixture.CreateTenureRequestObject))
+                .Then(t => _steps.ThenTheTenureDetailsAreReturnedAndIdIsNotEmpty(_tenureFixture))
+                .BDDfy();
+        }
```
Comment (mikemoody-amido) [inline TenureInformationApi.Tests/V1/E2ETests/Stories/PostTenureTests.cs:55]: Any failure cases? (There may not be yet...)

File: TenureInformationApi/V1/Boundary/Requests/CreateTenureRequestObject.cs (hunk 4 above)
Comment (humulla) [reply TenureInformationApi/V1/Boundary/Requests/CreateTenureRequestObject.cs:9]: I don't see any mention of it, but if anyone mentions any validation then I'm happy to add it in

File: TenureInformationApi/V1/Factories/ResponseFactory.cs (hunk 2 above)
Comment (humulla) [reply TenureInformationApi/V1/Factories/ResponseFactory.cs:8]: I've been able to get it working without injecting the DI container

File: TenureInformationApi.Tests/V1/E2ETests/Stories/PostTenureTests.cs (hunk 6 above)
Comment (humulla) [reply TenureInformationApi.Tests/V1/E2ETests/Stories/PostTenureTests.cs:55]: I guess validation check but that would be if we need to add any in

File: TenureInformationApi/V1/Boundary/Requests/CreateTenureRequestObject.cs (hunk 4 above)
Comment (humulla) [reply TenureInformationApi/V1/Boundary/Requests/CreateTenureRequestObject.cs:9]: I've added Validation and relevant tests

File: TenureInformationApi.Tests/V1/E2ETests/Stories/PostTenureTests.cs (hunk 6 above)
Comment (humulla) [reply TenureInformationApi.Tests/V1/E2ETests/Stories/PostTenureTests.cs:55]: I've added a failure case

File: TenureInformationApi/V1/Boundary/Requests/CreateTenureRequestObject.cs (hunk 3 above)
Comment (humulla) [reply TenureInformationApi/V1/Boundary/Requests/CreateTenureRequestObject.cs]: I've removed IsActive as we don't need it

File: TenureInformationApi/V1/Boundary/Requests/Validation/CreateTenureRequestValidation.cs [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+            RuleFor(x => x.Terminated).SetValidator(new TerminatedValidator());
+
+
+
+
+
+
+
+
+
```
Comment (mikemoody-amido) [inline TenureInformationApi/V1/Boundary/Requests/Validation/CreateTenureRequestValidation.cs:39]: Should remove

**Review by mikemoody-amido — APPROVED** [review]

**Pull request merged**